scripts/training/finetune-ldm-palette-adapter.py

```diff
@@ -0,0 +1,412 @@
+from PIL import Image
+from pydantic import BaseModel
+from torch import Tensor, device as Device, dtype as DType
+from torch.distributions import Beta
+from torch.nn.functional import mse_loss
+from torch.utils.data import Dataset
+from torchvision.transforms import Compose, RandomCrop, RandomHorizontalFlip
+
+import refiners.fluxion.layers as fl
+from refiners.fluxion.utils import save_to_safetensors
+from refiners.foundationals.clip.text_encoder import CLIPTextEncoderL
+from refiners.foundationals.latent_diffusion.palette_adapter import ColorEncoder, PaletteAdapter
+from refiners.foundationals.latent_diffusion.schedulers.ddpm import DDPM
+from refiners.foundationals.latent_diffusion.stable_diffusion_1.model import SD1Autoencoder
+from refiners.foundationals.latent_diffusion.stable_diffusion_1.unet import SD1UNet
+from refiners.training_utils.callback import Callback
+from refiners.training_utils.config import BaseConfig
+from refiners.training_utils.latent_diffusion import (
+    LatentDiffusionConfig,
+    TestDiffusionConfig,
+    resize_image,
+    sample_noise,
+)
+from refiners.training_utils.trainer import Trainer
+
+Image.MAX_IMAGE_PIXELS = None
+
+
+class ColorEncoderConfig(BaseModel):
+    """Configuration for the color encoder (of the palette adapter)."""
+
+    dim_sinusoids: int = 64
+    dim_embeddings: int = 256
+    max_colors: int = 8
+
+
+class AdapterConfig(BaseModel):
+    """Configuration for the palette adapter."""
+
+    color_encoder: ColorEncoderConfig
+
+
+class DatasetConfig(BaseModel):
+    """Configuration for the dataset."""
+
+    hf_repo: str
+    revision: str = "main"
+    split: str = "train"
+    horizontal_flip_probability: float = 0.5
+    resize_image_min_size: int = 512
+    resize_image_max_size: int = 576
+    random_crop_size: int = 512
+
+
+class TestPaletteDiffusionConfig(TestDiffusionConfig):
+    """Configuration to test the diffusion model, during the `evaluation` loop of the trainer."""
+
+    palettes: list[list[tuple[int, int, int]]] = []
+
+
+class AdapterLatentDiffusionConfig(BaseConfig):
+    """Finetunning configuration.
+
+    Contains the configs of the dataset, the latent diffusion model and the adapter.
+    """
+
+    dataset: DatasetConfig
+    ldm: LatentDiffusionConfig
+    adapter: AdapterConfig
+    test_ldm: TestPaletteDiffusionConfig
+
+
+class PaletteBatch(TypedDict):
+    """Structure of the data in the PaletteDataset."""
+
+    latent: Tensor
+    text_embedding: Tensor
+    palettes: dict[str, Tensor]
+
+
+class PaletteDataset(Dataset[PaletteBatch]):
+    """Dataset for the palette adapter.
+
+    Transforms the data from the Hugging Face dataset into PaletteBatches.
+    """
+
+    def __init__(self, trainer: "AdapterLatentDiffusionTrainer") -> None:
+        super().__init__()
+        self.trainer = trainer
+        self.dataset = self.load_huggingface_dataset()
+
+    @staticmethod
+    def download_image(
+        data: dict[str, list[Any]],  # "batched"
```

Comment:
Or `download_images` since it's batched (same applies below)

Reply:
Also, how does DownloadManager handles download failures? Is there a retry mechanism or something?

Reply:
[`DownloadManager.download()`](https://huggingface.co/docs/datasets/v2.16.1/en/package_reference/builder_classes#datasets.DownloadManager.download) calls at one point [`cached_path`](https://github.com/huggingface/datasets/blob/7b2bcd76457de720454c3ac304f2ed5c6f40acaa/src/datasets/utils/file_utils.py#L151) which raises ConnectionErrors. It looks like a `max_retries` (defaults to 1) (and a bunch of other options) can be configured through a [`DownloadConfig`](https://huggingface.co/docs/datasets/v2.16.1/en/package_reference/builder_classes#datasets.DownloadConfig).